File: tradingcards-plugin/src/main/java/net/tinetwork/tradingcards/tradingcardsplugin/commands/BuyCommand.java
package net.tinetwork.tradingcards.tradingcardsplugin.commands;

import co.aikar.commands.BaseCommand;
import co.aikar.commands.annotation.*;
import net.tinetwork.tradingcards.api.economy.ResponseWrapper;
import net.tinetwork.tradingcards.api.model.Rarity;
import net.tinetwork.tradingcards.api.model.Series;
import net.tinetwork.tradingcards.api.model.pack.Pack;
import net.tinetwork.tradingcards.api.model.pack.PackEntry;
import net.tinetwork.tradingcards.tradingcardsplugin.messages.internal.Permissions;
import net.tinetwork.tradingcards.tradingcardsplugin.TradingCards;
import net.tinetwork.tradingcards.tradingcardsplugin.card.EmptyCard;
import net.tinetwork.tradingcards.tradingcardsplugin.card.TradingCard;
import net.tinetwork.tradingcards.tradingcardsplugin.utils.CardUtil;
import net.tinetwork.tradingcards.tradingcardsplugin.utils.PlaceholderUtil;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * @author sarhatabaot
 */
@CommandAlias("cards")
public class BuyCommand extends BaseCommand {
    private final TradingCards plugin;

    public BuyCommand(final TradingCards plugin) {
        this.plugin = plugin;
    }

    @Subcommand("buy")
    @CommandPermission(Permissions.User.Economy.BUY)
    public class BuySubCommand extends BaseCommand {

        @Subcommand("pack")
        @CommandPermission(Permissions.User.Economy.BUY_PACK)
        @CommandCompletion("@packs")
        @Description("Buy a pack.")
        public void onBuyPack(final Player player, final String packId,final @Optional Integer amount) {
            if (CardUtil.noEconomy(player))
                return;

            if (plugin.getPackManager().getPack(packId) == null) {
                player.sendMessage(plugin.getMessagesConfig().packDoesntExist());
                return;
            }

            Pack pack = plugin.getPackManager().getPack(packId);

            if (pack.getBuyPrice() <= 0.0D && pack.getTradeCards().isEmpty()) {
                player.sendMessage(plugin.getMessagesConfig().cannotBeBought());
                return;
            }
    
            final int amountToBuy = getAmount(amount);
            
            if (!pack.getTradeCards().isEmpty() && !CardUtil.hasCardsInInventory(player, pack.getTradeCards())) {
                player.sendMessage("Not enough cards to perform a trade in.");
                return;
            }

            final double totalPrice = pack.getBuyPrice() * amountToBuy;
            ResponseWrapper economyResponse = plugin.getEconomyWrapper().withdraw(player, pack.getCurrencyId(), totalPrice);
            if (economyResponse.success()) {

                if (plugin.getGeneralConfig().closedEconomy()) {
                    plugin.getEconomyWrapper().depositAccount(plugin.getGeneralConfig().serverAccount(), pack.getCurrencyId(), totalPrice);
                }

                if (!pack.getTradeCards().isEmpty()) {
                    for (int i = 0; i <= amountToBuy; i++) {
                        tradeCards(pack, player);
                    }
                }

                player.sendMessage(plugin.getMessagesConfig().boughtCard().replaceAll(PlaceholderUtil.AMOUNT.asRegex(), String.valueOf(totalPrice)));
                
                final ItemStack packItem = plugin.getPackManager().getPackItem(packId);
                packItem.setAmount(amountToBuy);
                CardUtil.dropItem(player, packItem);
                return;
            }

            player.sendMessage(plugin.getMessagesConfig().notEnoughMoney());
        }
        
        private void tradeCards(final @NotNull Pack pack, final Player player) {
            final String packId = pack.getId();
            Map<ItemStack, Integer> removedCardsMap = new HashMap<>();
            for (PackEntry packEntry : pack.getTradeCards()) {
                Map<ItemStack, Integer> removedEntryItems = CardUtil.removeCardsMatchingEntry(player, packEntry);
        
                removedCardsMap = Stream.concat(removedCardsMap.entrySet().stream(), removedEntryItems.entrySet().stream()).collect(
                    Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
            }
            final int totalCards = removedCardsMap.values().stream().mapToInt(Integer::intValue).sum();
            player.sendMessage("Traded %s cards for pack %s:".formatted(totalCards,packId));
            CardUtil.sendTradedCardsMessage(player, removedCardsMap);
        }
        
        private int getAmount(final Integer amount) {
            if (amount == null || amount == 0) {
                return 1;
            }
            if (amount < 0) {
               return amount * -1;
            }
            return amount;
        }



        @Subcommand("card")
        @CommandPermission(Permissions.User.Economy.BUY_CARD)
        @Description("Buy a card.")
        @CommandCompletion("@rarities @series @cards")
        public void onBuyCard(final Player player, @NotNull final Rarity rarity, @NotNull final Series series, @NotNull final String cardId) {
            if (CardUtil.noEconomy(player))
                return;

            if (plugin.getCardManager().getCard(cardId, rarity.getId(), series.getId()) instanceof EmptyCard) {
                player.sendMessage(plugin.getMessagesConfig().cardDoesntExist());
                return;
            }

            final TradingCard tradingCard = plugin.getCardManager().getCard(cardId, rarity.getId(), series.getId());
            final double buyPrice = tradingCard.getBuyPrice();
            final String currencyId = tradingCard.getCurrencyId();

            ResponseWrapper economyResponse = plugin.getEconomyWrapper().withdraw(player, currencyId, buyPrice);
            if (economyResponse.success()) {
                if (plugin.getGeneralConfig().closedEconomy()) {
                    plugin.getEconomyWrapper().depositAccount(plugin.getGeneralConfig().serverAccount(), currencyId, buyPrice);
                }
                CardUtil.dropItem(player, tradingCard.build(false));
                player.sendMessage(plugin.getMessagesConfig().boughtCard().replaceAll(PlaceholderUtil.AMOUNT.asRegex(), String.valueOf(buyPrice)));
                return;
            }
            player.sendMessage(plugin.getMessagesConfig().notEnoughMoney());
        }
    }
}

File: tradingcards-plugin/src/main/kotlin/com/github/sarhatabaot/tradingcards/commands/BuyCommand.kt
package com.github.sarhatabaot.tradingcards.commands

import co.aikar.commands.BaseCommand
import co.aikar.commands.annotation.*
import com.github.sarhatabaot.tradingcards.card.EmptyCard
import com.github.sarhatabaot.tradingcards.extensions.hasCardsInInventory
import com.github.sarhatabaot.tradingcards.extensions.sendColoredMessage
import com.github.sarhatabaot.tradingcards.extensions.sendTradedCardsMessage
import com.github.sarhatabaot.tradingcards.utils.PlaceholderUtil
import net.tinetwork.tradingcards.api.manager.PackManager
import net.tinetwork.tradingcards.api.model.Rarity
import net.tinetwork.tradingcards.api.model.Series
import net.tinetwork.tradingcards.api.model.pack.Pack
import net.tinetwork.tradingcards.tradingcardsplugin.TradingCards
import net.tinetwork.tradingcards.tradingcardsplugin.managers.cards.AllCardManager
import net.tinetwork.tradingcards.tradingcardsplugin.messages.internal.Permissions
import net.tinetwork.tradingcards.tradingcardsplugin.utils.CardUtil
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.Stream


/**
 * @author sarhatabaot
 */
@CommandAlias("cards buy")
@CommandPermission(Permissions.User.Economy.BUY)
class BuyCommand(var plugin: TradingCards): BaseCommand() {
    private var packManager: PackManager = plugin.packManager
    private var cardManager: AllCardManager = plugin.cardManager

    @Subcommand("pack")
    @CommandPermission(Permissions.User.Economy.BUY_PACK)
    @CommandCompletion("@packs")
    @Description("Buy a pack.")
    fun onBuyPack(player: Player, packId: String, @Optional amount: Int) {
        if (CardUtil.noEconomy(player))
            return

        if (!packManager.containsPack(packId)) {
            player.sendMessage(plugin.messagesConfig.packDoesntExist())
            return
        }

        val pack = packManager.getPack(packId)
        if (pack.buyPrice <= 0.0 && pack.tradeCards.isEmpty()) {
            player.sendMessage(plugin.messagesConfig.cannotBeBought())
            return
        }
        val amountToBuy = getAmount(amount)
        if (pack.tradeCards.isNotEmpty() && !player.hasCardsInInventory(pack.tradeCards)) {
            player.sendMessage("Not enough cards to perform a trade in.")
            return
        }
        val totalPrice = pack.buyPrice * amountToBuy
        val economyResponse = plugin.economyWrapper.withdraw(player, pack.currencyId, totalPrice)
        if (economyResponse.success()) {
            if (plugin.generalConfig.closedEconomy()) {
                plugin.economyWrapper.depositAccount(plugin.generalConfig.serverAccount(), pack.currencyId, totalPrice)
            }
            if (pack.tradeCards.isNotEmpty()) {
                for (i in 0..amount) {
                    tradeCards(pack, player)
                }
            }
            player.sendMessage(
                plugin.messagesConfig.boughtCard()
                    .replace(PlaceholderUtil.AMOUNT.asRegex().toRegex(), totalPrice.toString())
            )
            val packItem = plugin.packManager.getPackItem(packId)
            packItem.amount = amount
            CardUtil.dropItem(player, packItem)
            return
        }
        player.sendMessage(plugin.messagesConfig.notEnoughMoney())
    }

    private fun tradeCards(pack: Pack, player: Player) {
        val packId = pack.id
        var removedCardsMap: Map<ItemStack, Int> = HashMap()
        for (packEntry in pack.tradeCards) {
            val removedEntryItems = CardUtil.removeCardsMatchingEntry(player, packEntry)
            removedCardsMap = Stream.concat<Map.Entry<ItemStack, Int>>(
                removedCardsMap.entries.stream(),
                removedEntryItems.entries.stream()
            ).collect(
                Collectors.toMap<Map.Entry<ItemStack, Int>, ItemStack, Int>(
                    Function<Map.Entry<ItemStack, Int>, ItemStack> { it: Map.Entry<ItemStack?, Int>? -> java.util.Map.Entry.key },
                    Function<Map.Entry<ItemStack, Int>, Int> { java.util.Map.Entry.value })
            )
        }
        val totalCards = removedCardsMap.values.stream().mapToInt { obj: Int -> obj }.sum()
        player.sendMessage("Traded $totalCards cards for pack $packId:")
        player.sendTradedCardsMessage(removedCardsMap)
    }

    private fun getAmount(amount: Int): Int {
        if (amount == 0) {
            return 1
        }
        return if (amount < 0) {
            amount * -1
        } else amount
    }


    @Subcommand("card")
    @CommandPermission(Permissions.User.Economy.BUY_CARD)
    @Description("Buy a card.")
    @CommandCompletion("@rarities @series @cards")
    fun onBuyCard(player: Player, rarity: Rarity, series: Series, cardId: String) {
        if (CardUtil.noEconomy(player)) return
        if (cardManager.getCard(cardId, rarity.id, series.id) is EmptyCard) {
            player.sendMessage(plugin.messagesConfig.cardDoesntExist())
            return
        }
        val tradingCard = cardManager.getCard(cardId, rarity.id, series.id)
        val buyPrice = tradingCard.buyPrice
        val currencyId = tradingCard.currencyId
        val economyResponse = plugin.economyWrapper.withdraw(player, currencyId, buyPrice)
        if (economyResponse.success()) {
            if (plugin.generalConfig.closedEconomy()) {
                plugin.economyWrapper.depositAccount(plugin.generalConfig.serverAccount(), currencyId, buyPrice)
            }
            CardUtil.dropItem(player, tradingCard.build(false))
            player.sendColoredMessage(
                plugin.messagesConfig.boughtCard()
                    .replace(PlaceholderUtil.AMOUNT.asRegex().toRegex(), buyPrice.toString())
            )
            return
        }
        player.sendColoredMessage(plugin.messagesConfig.notEnoughMoney())
    }
}